src/subgraph/service.ts

```diff
@@ -0,0 +1,822 @@
+    this.headers = this.buildHeaders(subgraphConfig.apiKey)
+  }
+
+  /**
+   * Resolves the GraphQL endpoint from configuration
+   */
+  private resolveEndpoint (config: SubgraphConfig): string {
+    if (config.endpoint != null && config.endpoint.trim() !== '') {
+      return config.endpoint.trim()
+    }
+
+    if (config.goldsky != null) {
+      return this.buildGoldskyEndpoint(config.goldsky)
+    }
+
+    throw createError(
+      'SubgraphService',
+      'constructor',
+      'Invalid configuration: provide either endpoint or complete goldsky config'
+    )
+  }
+
+  /**
+   * Builds Goldsky endpoint URL
+   */
+  private buildGoldskyEndpoint (goldsky: NonNullable<SubgraphConfig['goldsky']>): string {
+    const { projectId, subgraphName, version } = goldsky
+
+    if (
+      projectId?.trim() == null ||
+      projectId?.trim() === '' ||
+      subgraphName?.trim() == null ||
+      subgraphName?.trim() === '' ||
+      version?.trim() == null ||
+      version?.trim() === ''
+    ) {
+      throw createError(
+        'SubgraphService',
+        'constructor',
+        'Incomplete Goldsky config: projectId, subgraphName, and version required'
+      )
+    }
+
+    return `https://api.goldsky.com/api/public/${projectId}/subgraphs/${subgraphName}/${version}/gn`
+  }
+
+  /**
+   * Builds HTTP headers for requests
+   */
+  private buildHeaders (apiKey?: string): Record<string, string> {
+    const headers = { 'Content-Type': 'application/json' }
+
+    if (apiKey != null && apiKey !== '') {
+      return { ...headers, Authorization: `Bearer ${apiKey}` }
+    }
+
+    return headers
+  }
+
+  /**
+   * Executes a GraphQL query
+   */
+  private async executeQuery<T>(
+    query: string,
+    variables: Record<string, any>,
+    operation: string
+  ): Promise<T> {
+    try {
+      const response = await fetch(this.endpoint, {
+        method: 'POST',
+        headers: this.headers,
+        body: JSON.stringify({ query, variables })
+      })
+
+      if (!response.ok) {
+        const errorText = await response.text()
+        throw createError('SubgraphService', operation, `HTTP ${response.status}: ${errorText}`)
+      }
+
+      const result = (await response.json()) as GraphQLResponse<T>
+
+      if (result.errors != null && result.errors.length > 0) {
+        const errorMsg = result.errors.map((e) => e.message).join('; ')
+        throw createError('SubgraphService', operation, `GraphQL errors: ${errorMsg}`)
+      }
+
+      return result.data as T
+    } catch (error) {
+      if (error instanceof Error && error.name === 'SynapseError') {
+        throw error
+      }
+
+      throw createError(
+        'SubgraphService',
+        operation,
+        `Query execution failed: ${(error as Error).message}`,
+        { cause: error }
+      )
+    }
+  }
+
+  /**
+   * Converts CommP to hex string
+   */
+  private commPToHex (commP: CommP): string {
+    return toHex(CID.parse(commP.toString()).bytes)
+  }
+
```

Comment:
you have a `CommP` here, which is a type of `CID` so you're doing a cid->string->parse operation here that's most likely going to be useless because it's already done; but you can do a quick/cheap validation/parse with `asCommP` from commp.ts that ends up being a noop if it's already what you need:

```suggestion
  private commPToHex (commP: CommP): string {
    const parsedCommP = asCommP(commP)
    if (parsedCommP == null) {
      throw new Error(`Invalid CommP: ${String(expectedCommP)}`)
    }
    return toHex(parsedCommP.bytes)
  }
```